File: packages/vscode-extension/src/extension/session/indicators.ts
import { type Uri, workspace } from "#vscode-host";
import type { NativeSuggestionIndicators } from "../native/config.ts";
import { configuredValue } from "./configured.ts";

const suggestionIndicatorKeys = [
  ["build", "Build"],
  ["latest", "Latest"],
  ["satisfiesLatest", "SatisfiesLatest"],
  ["directory", "Directory"],
  ["error", "Error"],
  ["matched", "Match"],
  ["noMatch", "NoMatch"],
  ["updateable", "Updateable"],
  ["updateableVulnerable", "UpdateableVulnerable"],
] as const;

type LegacySuggestionIndicators = Readonly<Record<string, string | undefined>>;

export function suggestionIndicators(
  resource?: Uri,
): NativeSuggestionIndicators | undefined {
  const config = workspace.getConfiguration("versionlens", resource);
  const configuredNative = configuredValue<
    NativeSuggestionIndicators | undefined
  >("suggestions.indicators", config);
  const configuredLegacy = configuredValue<
    LegacySuggestionIndicators | undefined
  >("suggestions.indicators", config);
  if (!(configuredNative || configuredLegacy)) {
    return;
  }

  const indicators: NativeSuggestionIndicators = {};
  for (const [nativeKey, legacyKey] of suggestionIndicatorKeys) {
    const value =
      configuredNative?.[nativeKey] ?? configuredLegacy?.[legacyKey];
    if (value !== undefined) {
      indicators[nativeKey] = value;
    }
  }
  return indicators;
}

File: crates/versionlens-napi/src/model/config/session/tests/indicators.rs
use super::blank_session_config;
use crate::model::config::NativeSessionConfig;
use crate::model::config::NativeSuggestionIndicators;

fn blank_indicators() -> NativeSuggestionIndicators {
    NativeSuggestionIndicators {
        latest: None,
        satisfies_latest: None,
        directory: None,
        error: None,
        no_match: None,
        matched: None,
        updateable: None,
        updateable_vulnerable: None,
        build: None,
    }
}

#[test]
fn partial_indicator_config_keeps_core_defaults() {
    let config = NativeSessionConfig {
        suggestion_indicators: Some(NativeSuggestionIndicators {
            updateable: Some("U".to_owned()),
            ..blank_indicators()
        }),
        ..blank_session_config()
    }
    .into_core();

    assert_eq!(config.suggestion_indicators.latest, "\u{1F7E2}");
    assert_eq!(config.suggestion_indicators.updateable, "U");
    assert_eq!(config.suggestion_indicators.directory, "\u{1F4C1}");
    assert_eq!(config.suggestion_indicators.error, "\u{1F534}");
    assert_eq!(config.suggestion_indicators.matched, "\u{1F7E1}");
    assert_eq!(config.suggestion_indicators.build, "\u{224C} ");
    assert_eq!(
        config.suggestion_indicators.updateable_vulnerable,
        "\u{26A0}\u{FE0F}"
    );
}

#[test]
fn directory_indicator_maps_to_core() {
    let config = NativeSessionConfig {
        suggestion_indicators: Some(NativeSuggestionIndicators {
            directory: Some("D".to_owned()),
            ..blank_indicators()
        }),
        ..blank_session_config()
    }
    .into_core();

    assert_eq!(config.suggestion_indicators.directory, "D");
}
